### ISSUE TITLE
DocCommentAlignmentSniff does not format Doctrine's nested annotations properly

### ISSUE THREAD

**Issue by sustmi** [squizlabs/PHP_CodeSniffer#1936]
I created a (failing) test case that describes the problem here: https://github.com/squizlabs/PHP_CodeSniffer/compare/master...sustmi:doctrine-nested-annotations-test-case?expand=1

This is the output from PHPUnit:
```diff
There was 1 failure:

1) PHP_CodeSniffer\Standards\Squiz\Tests\Commenting\DocCommentAlignmentUnitTest::testSniff
[LINE 90] Expected 0 error(s) in DocCommentAlignmentUnitTest.inc but found 1 error(s). The error(s) found were:
 -> Expected 1 space after asterisk; 9 found (Squiz.Commenting.DocCommentAlignment.SpaceAfterStar)
[LINE 91] Expected 0 error(s) in DocCommentAlignmentUnitTest.inc but found 1 error(s). The error(s) found were:
 -> Expected 1 space after asterisk; 9 found (Squiz.Commenting.DocCommentAlignment.SpaceAfterStar)
Fixed version of DocCommentAlignmentUnitTest.inc does not match expected version in DocCommentAlignmentUnitTest.inc.fixed; the diff is
--- src/Standards/Squiz/Tests/Commenting/DocCommentAlignmentUnitTest.inc.fixed
+++ PHP_CodeSniffer
@@ -87,8 +87,8 @@
  * @ORM\Table(
  *     name="table_name",
  *     indexes={
- *         @ORM\Index(columns={"my_column"}),
- *         @ORM\Index(columns={"my_column2"})
+ * @ORM\Index(columns={"my_column"}),
+ * @ORM\Index(columns={"my_column2"})
  *     }
  * )
  */
```

**Comment by sustmi** [comment]
I guess that in order to process the Doctrine annotations properly it would be necessary to use https://github.com/doctrine/annotations/blob/master/lib/Doctrine/Common/Annotations/DocParser.php .

I wonder whether there is some other way/workaround how to keep Doctrine anontation's arguments indented.

**Comment by sustmi** [comment]
Oh, I can see that CodeSniffer has it's own comment tokenizer https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Tokenizers/Comment.php .

Maybe we could extend it to support tokenizing Doctrine-style arguments of annotation tags. Then we could easily ignore all code in `@Tag(...)`. (Or we could even possibly extend the sniff to properly indent all the arguments.)

**Comment by gsherwood** [comment]
Yes, this would need a custom sniff unless this syntax applies to more than just Doctrine, which I don't think it does.

You could certainly try writing a custom sniff using the built-in comment tokenizer like the existing DocCommentAlignment sniff does. For the doc comment you linked to:
```php
/**
 * @ORM\Entity()
 * @ORM\Table(
 *     name="table_name",
 *     indexes={
 *         @ORM\Index(columns={"my_column"}),
 *         @ORM\Index(columns={"my_column2"})
 *     }
 * )
 */
class MyClass
{

}
```
You would get these tokens:
```
Create comment token: T_DOC_COMMENT_OPEN_TAG => /**
Create comment token: T_DOC_COMMENT_WHITESPACE => \n
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_STAR => *
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_TAG => @ORM\Entity()
Create comment token: T_DOC_COMMENT_WHITESPACE => \n
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_STAR => *
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_TAG => @ORM\Table(
Create comment token: T_DOC_COMMENT_WHITESPACE => \n
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_STAR => *
Create comment token: T_DOC_COMMENT_WHITESPACE => ·····
Create comment token: T_DOC_COMMENT_STRING => name="table_name",
Create comment token: T_DOC_COMMENT_WHITESPACE => \n
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_STAR => *
Create comment token: T_DOC_COMMENT_WHITESPACE => ·····
Create comment token: T_DOC_COMMENT_STRING => indexes={
Create comment token: T_DOC_COMMENT_WHITESPACE => \n
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_STAR => *
Create comment token: T_DOC_COMMENT_WHITESPACE => ·········
Create comment token: T_DOC_COMMENT_TAG => @ORM\Index(columns={"my_column"}),
Create comment token: T_DOC_COMMENT_WHITESPACE => \n
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_STAR => *
Create comment token: T_DOC_COMMENT_WHITESPACE => ·········
Create comment token: T_DOC_COMMENT_TAG => @ORM\Index(columns={"my_column2"})
Create comment token: T_DOC_COMMENT_WHITESPACE => \n
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_STAR => *
Create comment token: T_DOC_COMMENT_WHITESPACE => ·····
Create comment token: T_DOC_COMMENT_STRING => }
Create comment token: T_DOC_COMMENT_WHITESPACE => \n
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_STAR => *
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_STRING => )
Create comment token: T_DOC_COMMENT_WHITESPACE => \n
Create comment token: T_DOC_COMMENT_WHITESPACE => ·
Create comment token: T_DOC_COMMENT_CLOSE_TAG => */
```
If you think you could look at those tokens and figure out the correct indentation, then a custom sniff should work fine.

**Comment by sustmi** [comment]
I found out that PHP CS Fixer also does Doctrine annotation tokenization (https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/v2.10.4/src/Doctrine/Annotation/Tokens.php#L34) and it uses Doctrine's [`DocLexer`](https://github.com/doctrine/annotations/blob/master/lib/Doctrine/Common/Annotations/DocLexer.php) for it and therefore it tokenizes Doctrine annotation arguments.

**Comment by sustmi** [comment]
After spending some time looking at [`\PHP_CodeSniffer\Tokenizers\Comment`](https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Tokenizers/Comment.php) on one side and [`\Doctrine\Common\Annotations\DocLexer`](https://github.com/doctrine/annotations/blob/master/lib/Doctrine/Common/Annotations/DocLexer.php) + [\PhpCsFixer\Doctrine\Annotation\Tokens](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.10/src/Doctrine/Annotation/Tokens.php) on the other side, I have two thoughts in my mind: *reinventing the wheel* and *not worth it*.

I use both *CodeSniffer* and *PHP CS Fixer* to check/fix my code and *PHP CS Fixer* has sufficient support for Doctrine annotations for me (for more details search for `doctrine_annotation` here: https://github.com/FriendsOfPHP/PHP-CS-Fixer). I do not need two tools doing the same thing so I do not really expect *CodeSniffer* to check/fix Doctrine annotations. I just want it to not report false errors.

**Therefore I suggest the following solution/workaround:** ignore all Doctrine annotations in `DocCommentAlignmentSniff`. To be concrete, if `T_DOC_COMMENT_TAG` token contains an opening brace (eg. `@ORM\Table(...`) it will not enforce the "Expected 1 space after asterisk" rule.

This way it will pass all the test cases (including the one mentioned in the description of the issue) and I will leave the proper checking/fixing of Doctrine annotations to *PHP CS Fixer* in my projects.

What do you think @gsherwood ?

**Comment by gsherwood** [comment]
> What do you think @gsherwood ?

I'm not keen on coding in exceptions for specific frameworks in any sniffs.

If I made this change, I can't guarantee that the sniff will continue to report indentation errors for code that it should. I can't make a blanket rule that tags with parenthesis must be doctrine tags and so they should be ignored.

I'm also not that keen on adding config options for specific frameworks, so I don't want to make this optional.

The best way to enforce different rules for different frameworks is to write different sniffs for them. Maybe you could talk whoever maintains https://github.com/doctrine/coding-standard and see if they would accept a custom sniff based on (or extending) DocCommentAlignment to properly check Doctrine comments.

**Comment by sustmi** [comment]
Meanwhile I made a try to implement the workaround that I described: https://github.com/squizlabs/PHP_CodeSniffer/compare/master...sustmi:doctrine-annotations-are-not-doc-comment-tags?expand=1. However it turned out to be just an exercise. :)

I understand your arguments and I just realized that I can solve my problem just by disabling `DocCommentAlignmentSniff` in my ruleset because the sniff does not have any extra value over *PHP CS Fixer*'s fixes for me. I do not know why I was fixed to the idea that the `DocCommentAlignmentSniff` has to play well with Doctrine annotations.

I guess we can close the issue because it **does not need to be fixed**. Thank you for helping me.
